storage/database/db_manager.go

```diff
@@ -371,10 +387,85 @@ func (dbm *databaseManager) IsParallelDBWrite() bool {
 	return dbm.config.ParallelDBWrite
 }
 
+func (dbm *databaseManager) InMigration() bool {
+	return dbm.inMigration
+}
+
 func (dbm *databaseManager) NewBatch(dbEntryType DBEntryType) Batch {
+	if dbEntryType == StateTrieDB && dbm.inMigration {
+		return dbm.newStateTrieDB.NewBatch()
+	}
 	return dbm.getDatabase(dbEntryType).NewBatch()
 }
 
+func (dbm *databaseManager) getStateTrieMigrationInfo() (bool, string) {
+	miscDB := dbm.getDatabase(MiscDB)
+
+	enc, _ := miscDB.Get(migrationStatusKey)
+	if len(enc) == 0 {
+		return false, ""
+	}
+
+	var status uint64
+	if err := rlp.DecodeBytes(enc, &status); err != nil {
+		logger.Error("Failed to decode migration status", "err", err)
+		return false, ""
+	}
+
+	if status == 0 {
+		return false, ""
+	}
+
+	enc, _ = miscDB.Get(databaseDirKey(uint64(StateTrieDB)))
+	return true, string(enc)
+}
+
+func createStateTrieMigrationDB(dbc *DBConfig, newDBDir string) Database {
+	oldDBConfig := dbc
+	newDBConfig := getDBEntryConfig(oldDBConfig, StateTrieDB)
+	newDBConfig.Dir = newDBDir
+	var newDB Database
+	var err error
+	if newDBConfig.NumStateTriePartitions > 1 {
+		newDB, err = newPartitionedDB(newDBConfig, StateTrieDB, newDBConfig.NumStateTriePartitions)
+	} else {
+		newDB, err = newDatabase(newDBConfig, StateTrieDB)
+	}
+	if err != nil {
+		logger.Crit("Failed to create a new database for state trie migration", "err", err)
+	}
+	logger.Info("Created a new database for state trie migration", "newStateTrieDB", newDBConfig.Dir)
+	return newDB
+}
+
+func (dbm *databaseManager) SetStateTrieMigrationDB(blockNum uint64) {
+	if !dbm.config.Partitioned {
+		logger.Warn("Setting a new database for state trie migration is allowed for partitioned database only")
```

Comment:
It doesn't seems that only partitioned database is allow in createStateTrieMigrationDB.
Why this warn and return are needed?

Reply:
Migration of non-partitioned database does not seem needed and helpful. Maybe we can do that but I don't think it is needed. Since in that case, as there is only one LevelDB instance, read-request from migration goroutine will continuously block read/write operations from on-going side.